kernel/src/sched.rs

```diff
@@ -41,7 +68,7 @@ pub struct Kernel {
 }
 
 impl Kernel {
-    pub fn new(processes: &'static [Option<&'static process::ProcessType>]) -> Kernel {
+    pub fn new(processes: &'static Processes) -> Kernel {
```

Comment:
Is there a way to implement this that doesn't require changing this type? That would enable these changes to all be inside of the kernel crate, and not require updating all boards.

Also, I think it would make it easier to add new scheduler policies without having to change the main.rs files (although there are probably ways to hide that if this type does change).

Reply:
Seems plausible... though...

Having a "special" `Processes" type would actually let us make other changes in the future without requiring changes to `main.rs`, so long as `Processes` is opaque (that is, boards don't get to inspect the structure at all, and would create one with a utility function or something).

Reply:
I think making a `Processes` like type is reasonable, but I also think that having the interface be "here is an array of processes I want you, the kernel, to execute, and if you need any other state then feel free to manage that yourself" makes sense.

The board will also need a way to specify the scheduling algorithm, which will cause the interface to change.

Reply:
Yeah, though making the kernel decide how to manage it completely transparently to the board means the kernel needs to figure out how to allocate space for it, so... And it’s not a list of processes anyway, it’s a list of process _slots_ (which is a memory allocation problem), some of which are already filled. It’s currently a fairly implementation specific interface :/

> The board will also need a way to specify the scheduling algorithm, which will cause the interface to change.

We could do this with a type variable with a default value, I think:

```rust
...
impl Kernel<S: Scheduler = RoundRobinScheduler> {
    fn new(...) -> Kernel<S> {

    }
} 
...
```

Such that if you don’t specify anything (and thus your interface hasn’t changed) you get a default behavior. I _think_... haven’t tried, but something like that is supposed to work.

Reply:
Do you think that since we do a `static_init!()` for the Kernel struct that would be sufficient for allocating space for any scheduling state?